File: backend/src/ZooSanMarino.Infrastructure/Services/SeguimientoLoteLevanteService.cs
using System.Linq.Expressions;
using Microsoft.EntityFrameworkCore;
using ZooSanMarino.Application.DTOs;
using ZooSanMarino.Application.Interfaces;
using ZooSanMarino.Domain.Entities;
using ZooSanMarino.Infrastructure.Persistence;

namespace ZooSanMarino.Infrastructure.Services;

public class SeguimientoLoteLevanteService : ISeguimientoLoteLevanteService
{
    private readonly ZooSanMarinoContext _ctx;
    private readonly IAlimentoNutricionProvider _alimentos;
    private readonly IGramajeProvider _gramaje;
    private readonly ICurrentUser _current;

    public SeguimientoLoteLevanteService(
        ZooSanMarinoContext ctx,
        IAlimentoNutricionProvider alimentos,
        IGramajeProvider gramaje,
        ICurrentUser current)
    {
        _ctx = ctx;
        _alimentos = alimentos;
        _gramaje = gramaje;
        _current = current;
    }

    // Mapeo a DTO actualizado con los nuevos campos opcionales
    private static readonly Expression<Func<SeguimientoLoteLevante, SeguimientoLoteLevanteDto>> ToDto =
        x => new SeguimientoLoteLevanteDto(
            x.Id, x.LoteId, x.FechaRegistro,
            x.MortalidadHembras, x.MortalidadMachos,
            x.SelH, x.SelM,
            x.ErrorSexajeHembras, x.ErrorSexajeMachos,
            x.ConsumoKgHembras, x.TipoAlimento, x.Observaciones,
            x.KcalAlH, x.ProtAlH, x.KcalAveH, x.ProtAveH, x.Ciclo,

            // NUEVOS
            x.ConsumoKgMachos,
            x.PesoPromH, x.PesoPromM,
            x.UniformidadH, x.UniformidadM,
            x.CvH, x.CvM
        );

    // ===========================
    // LISTAR POR LOTE (seguro por compañía)
    // ===========================
    public async Task<IEnumerable<SeguimientoLoteLevanteDto>> GetByLoteAsync(int loteId)
    {
        var q = from s in _ctx.SeguimientoLoteLevante.AsNoTracking()
                join l in _ctx.Lotes.AsNoTracking()
                    on s.LoteId equals l.LoteId
                where l.CompanyId == _current.CompanyId
                   && l.DeletedAt == null
                   && s.LoteId == loteId
                select s;

        return await q
            .OrderBy(x => x.FechaRegistro)
            .Select(ToDto)
            .ToListAsync();
    }

    // ===========================
    // CREAR
    // ===========================
    public async Task<SeguimientoLoteLevanteDto> CreateAsync(SeguimientoLoteLevanteDto dto)
    {
        // 0) Verificar lote y tenant
        var lote = await _ctx.Lotes.AsNoTracking()
            .SingleOrDefaultAsync(l => l.LoteId == dto.LoteId
                                     && l.CompanyId == _current.CompanyId
                                     && l.DeletedAt == null);
        if (lote is null) throw new InvalidOperationException($"Lote '{dto.LoteId}' no existe o no pertenece a la compañía.");

        // 1) Duplicado por Lote+Fecha
        var duplicado = await (from s in _ctx.SeguimientoLoteLevante
                               where s.LoteId == dto.LoteId
                                  && s.FechaRegistro.Date == dto.FechaRegistro.Date
                               select s.Id).AnyAsync();
        if (duplicado) throw new InvalidOperationException("Ya existe un registro para ese Lote y Fecha.");

        // 2) Autocompletar nutrientes si no vienen
        double? kcalAlH = dto.KcalAlH, protAlH = dto.ProtAlH;
        if (kcalAlH is null || protAlH is null)
        {
            var np = await _alimentos.GetNutrientesAsync(dto.TipoAlimento);
            if (np.HasValue)
            {
                kcalAlH ??= np.Value.kcal;
                protAlH ??= np.Value.prot;
            }
        }

        // 3) Sugerir consumo por gramaje si <= 0 (hembras)
        double consumoKgH = dto.ConsumoKgHembras;
        if (consumoKgH <= 0 && !string.IsNullOrWhiteSpace(lote.GalponId) && lote.FechaEncaset.HasValue)
        {
            int semana = CalcularSemana(lote.FechaEncaset.Value, dto.FechaRegistro);

            double? gramajeGrAve = null;
            if (int.TryParse(lote.GalponId, out var galponIdInt))
                gramajeGrAve = await _gramaje.GetGramajeGrPorAveAsync(galponIdInt, semana, dto.TipoAlimento);
            else
            {
                if (_gramaje is IGramajeProviderV2 v2)
                    gramajeGrAve = await v2.GetGramajeGrPorAveAsync(lote.GalponId, semana, dto.TipoAlimento);
            }

            if (gramajeGrAve.HasValue && gramajeGrAve.Value > 0)
            {
                int hembrasVivas = await CalcularHembrasVivasAsync(dto.LoteId);
                consumoKgH = Math.Round((gramajeGrAve.Value * hembrasVivas) / 1000.0, 3); // gr → kg
            }
        }

        // 4) Derivados (por ahora solo hembras, igual que antes)
        var (kcalAveH, protAveH) = CalcularDerivados(consumoKgH, kcalAlH, protAlH);

        var ent = new SeguimientoLoteLevante
        {
            LoteId = dto.LoteId,
            FechaRegistro = dto.FechaRegistro,
            MortalidadHembras = dto.MortalidadHembras,
            MortalidadMachos = dto.MortalidadMachos,
            SelH = dto.SelH,
            SelM = dto.SelM,
            ErrorSexajeHembras = dto.ErrorSexajeHembras,
            ErrorSexajeMachos = dto.ErrorSexajeMachos,

            ConsumoKgHembras = consumoKgH,
            ConsumoKgMachos = dto.ConsumoKgMachos,

            PesoPromH = dto.PesoPromH,
            PesoPromM = dto.PesoPromM,
            UniformidadH = dto.UniformidadH,
            UniformidadM = dto.UniformidadM,
            CvH = dto.CvH,
            CvM = dto.CvM,

            TipoAlimento = dto.TipoAlimento,
            Observaciones = dto.Observaciones,

            KcalAlH = kcalAlH,
            ProtAlH = protAlH,
            KcalAveH = kcalAveH,
            ProtAveH = protAveH,

            Ciclo = dto.Ciclo
        };

        _ctx.SeguimientoLoteLevante.Add(ent);
        await _ctx.SaveChangesAsync();

        return await _ctx.SeguimientoLoteLevante.AsNoTracking()
            .Where(x => x.Id == ent.Id)
            .Select(ToDto)
            .SingleAsync();
    }

    // ===========================
    // ACTUALIZAR
    // ===========================
    public async Task<SeguimientoLoteLevanteDto?> UpdateAsync(SeguimientoLoteLevanteDto dto)
    {
        var ent = await _ctx.SeguimientoLoteLevante.FindAsync(dto.Id);
        if (ent is null) return null;

        // Validar lote + tenant
        var lote = await _ctx.Lotes.AsNoTracking()
            .SingleOrDefaultAsync(l => l.LoteId == dto.LoteId
                                     && l.CompanyId == _current.CompanyId
                                     && l.DeletedAt == null);
        if (lote is null) throw new InvalidOperationException($"Lote '{dto.LoteId}' no existe o no pertenece a la compañía.");

        // Duplicado al cambiar Lote/Fecha
        var cambiaClave = ent.LoteId != dto.LoteId || ent.FechaRegistro.Date != dto.FechaRegistro.Date;
        if (cambiaClave)
        {
            var existeOtro = await _ctx.SeguimientoLoteLevante.AnyAsync(x =>
                x.Id != dto.Id && x.LoteId == dto.LoteId && x.FechaRegistro.Date == dto.FechaRegistro.Date);
            if (existeOtro) throw new InvalidOperationException("Ya existe un registro para ese Lote y Fecha.");
        }

        // Nutrientes (si faltan)
        double? kcalAlH = dto.KcalAlH, protAlH = dto.ProtAlH;
        if (kcalAlH is null || protAlH is null)
        {
            var np = await _alimentos.GetNutrientesAsync(dto.TipoAlimento);
            if (np.HasValue)
            {
                kcalAlH ??= np.Value.kcal;
                protAlH ??= np.Value.prot;
            }
        }

        // Consumo sugerido si no viene (>0 respeta valor del usuario)
        double consumoKgH = dto.ConsumoKgHembras;
        if (consumoKgH <= 0 && !string.IsNullOrWhiteSpace(lote.GalponId) && lote.FechaEncaset.HasValue)
        {
            int semana = CalcularSemana(lote.FechaEncaset.Value, dto.FechaRegistro);

            double? gramajeGrAve = null;
            if (int.TryParse(lote.GalponId, out var galponIdInt))
                gramajeGrAve = await _gramaje.GetGramajeGrPorAveAsync(galponIdInt, semana, dto.TipoAlimento);
            else if (_gramaje is IGramajeProviderV2 v2)
                gramajeGrAve = await v2.GetGramajeGrPorAveAsync(lote.GalponId, semana, dto.TipoAlimento);

            if (gramajeGrAve.HasValue && gramajeGrAve.Value > 0)
            {
                int hembrasVivas = await CalcularHembrasVivasAsync(dto.LoteId);
                consumoKgH = Math.Round((gramajeGrAve.Value * hembrasVivas) / 1000.0, 3);
            }
        }

        var (kcalAveH, protAveH) = CalcularDerivados(consumoKgH, kcalAlH, protAlH);

        ent.LoteId = dto.LoteId;
        ent.FechaRegistro = dto.FechaRegistro;
        ent.MortalidadHembras = dto.MortalidadHembras;
        ent.MortalidadMachos = dto.MortalidadMachos;
        ent.SelH = dto.SelH;
        ent.SelM = dto.SelM;
        ent.ErrorSexajeHembras = dto.ErrorSexajeHembras;
        ent.ErrorSexajeMachos = dto.ErrorSexajeMachos;

        ent.ConsumoKgHembras = consumoKgH;
        ent.ConsumoKgMachos = dto.ConsumoKgMachos;

        ent.PesoPromH = dto.PesoPromH;
        ent.PesoPromM = dto.PesoPromM;
        ent.UniformidadH = dto.UniformidadH;
        ent.UniformidadM = dto.UniformidadM;
        ent.CvH = dto.CvH;
        ent.CvM = dto.CvM;

        ent.TipoAlimento = dto.TipoAlimento;
        ent.Observaciones = dto.Observaciones;

        ent.KcalAlH = kcalAlH;
        ent.ProtAlH = protAlH;
        ent.KcalAveH = kcalAveH;
        ent.ProtAveH = protAveH;

        ent.Ciclo = dto.Ciclo;

        await _ctx.SaveChangesAsync();

        return new SeguimientoLoteLevanteDto(
            ent.Id, ent.LoteId, ent.FechaRegistro,
            ent.MortalidadHembras, ent.MortalidadMachos,
            ent.SelH, ent.SelM, ent.ErrorSexajeHembras, ent.ErrorSexajeMachos,
            ent.ConsumoKgHembras, ent.TipoAlimento, ent.Observaciones,
            ent.KcalAlH, ent.ProtAlH, ent.KcalAveH, ent.ProtAveH, ent.Ciclo,

            // NUEVOS
            ent.ConsumoKgMachos,
            ent.PesoPromH, ent.PesoPromM,
            ent.UniformidadH, ent.UniformidadM,
            ent.CvH, ent.CvM
        );
    }

    // ===========================
    // ELIMINAR
    // ===========================
    public async Task<bool> DeleteAsync(int id)
    {
        var ent = await _ctx.SeguimientoLoteLevante.FindAsync(id);
        if (ent is null) return false;

        // (Opcional) Validar que el lote pertenezca a la compañía actual
        var ok = await _ctx.Lotes.AsNoTracking()
            .AnyAsync(l => l.LoteId == ent.LoteId && l.CompanyId == _current.CompanyId);
        if (!ok) return false;

        _ctx.SeguimientoLoteLevante.Remove(ent);
        await _ctx.SaveChangesAsync();
        return true;
    }

    // ===========================
    // FILTRO (seguro por compañía)
    // ===========================
    public async Task<IEnumerable<SeguimientoLoteLevanteDto>> FilterAsync(int? loteId, DateTime? desde, DateTime? hasta)
    {
        var q = from s in _ctx.SeguimientoLoteLevante.AsNoTracking()
                join l in _ctx.Lotes.AsNoTracking()
                    on s.LoteId equals l.LoteId
                where l.CompanyId == _current.CompanyId && l.DeletedAt == null
                select s;

        if (loteId.HasValue)
            q = q.Where(x => x.LoteId == loteId.Value);
        if (desde.HasValue)
            q = q.Where(x => x.FechaRegistro >= desde.Value);
        if (hasta.HasValue)
            q = q.Where(x => x.FechaRegistro <= hasta.Value);

        return await q
            .OrderBy(x => x.LoteId).ThenBy(x => x.FechaRegistro)
            .Select(ToDto)
            .ToListAsync();
    }

    // ===========================
    // Helpers
    // ===========================
    private static (double? kcalAveH, double? protAveH) CalcularDerivados(double consumoKgHembras, double? kcalAlH, double? protAlH)
    {
        double? kcal = (kcalAlH is null) ? null : Math.Round(consumoKgHembras * kcalAlH.Value, 3);
        double? prot = (protAlH is null) ? null : Math.Round(consumoKgHembras * protAlH.Value, 3);
        return (kcal, prot);
    }

    private static int CalcularSemana(DateTime fechaEncaset, DateTime fechaRegistro)
    {
        var dias = (fechaRegistro.Date - fechaEncaset.Date).TotalDays;
        return Math.Max(1, (int)Math.Floor(dias / 7.0) + 1);
    }

    private async Task<int> CalcularHembrasVivasAsync(int loteId)
    {
        var baseH = await _ctx.Lotes.AsNoTracking()
            .Where(l => l.LoteId == loteId && l.CompanyId == _current.CompanyId && l.DeletedAt == null)
            .Select(l => new { Base = l.HembrasL ?? 0, MortCaja = l.MortCajaH ?? 0 })
            .SingleAsync();

        var sum = await _ctx.SeguimientoLoteLevante.AsNoTracking()
            .Where(x => x.LoteId == loteId)
            .GroupBy(_ => 1)
            .Select(g => new
            {
                MortH = g.Sum(x => x.MortalidadHembras),
                SelH = g.Sum(x => x.SelH),
                ErrH = g.Sum(x => x.ErrorSexajeHembras)
            })
            .SingleOrDefaultAsync();

        int mort = sum?.MortH ?? 0;
        int sel = sum?.SelH ?? 0;
        int err = sum?.ErrH ?? 0;

        var vivas = baseH.Base - baseH.MortCaja - mort - sel - err;
        return Math.Max(0, vivas);
    }
    
      public async Task<ResultadoLevanteResponse> GetResultadoAsync(
        int loteId, DateTime? desde, DateTime? hasta, bool recalcular = true)
    {
        // 0) Validar lote/tenant
        var lote = await _ctx.Lotes.AsNoTracking()
            .SingleOrDefaultAsync(l => l.LoteId == loteId
                                     && l.CompanyId == _current.CompanyId
                                     && l.DeletedAt == null);
        if (lote is null)
            throw new InvalidOperationException($"Lote '{loteId}' no existe o no pertenece a la compañía.");

        // 1) Recalcular on-demand (opcional)
        if (recalcular)
        {
            // IMPORTANTE: usa parámetro para evitar SQL injection
            await _ctx.Database.ExecuteSqlInterpolatedAsync(
                $"select sp_recalcular_seguimiento_levante({loteId})");
        }

        // 2) Query del snapshot
        var q = from r in _ctx.ProduccionResultadoLevante.AsNoTracking()
                where r.LoteId == loteId
                select r;

        if (desde.HasValue) q = q.Where(x => x.Fecha >= desde.Value.Date);
        if (hasta.HasValue) q = q.Where(x => x.Fecha <= hasta.Value.Date);

        var items = await q.OrderBy(x => x.Fecha)
            .Select(r => new ResultadoLevanteItemDto(
                r.Fecha, r.EdadSemana,
                // H
                r.HembraViva, r.MortH, r.SelHOut, r.ErrH,
                r.ConsKgH, r.PesoH, r.UnifH, r.CvH,
                r.MortHPct, r.SelHPct, r.ErrHPct,
                r.MsEhH, r.AcMortH, r.AcSelH, r.AcErrH,
                r.AcConsKgH, r.ConsAcGrH, r.GrAveDiaH,
                r.DifConsHPct, r.DifPesoHPct, r.RetiroHPct, r.RetiroHAcPct,
                // M
                r.MachoVivo, r.MortM, r.SelMOut, r.ErrM,
                r.ConsKgM, r.PesoM, r.UnifM, r.CvM,
                r.MortMPct, r.SelMPct, r.ErrMPct,
                r.MsEmM, r.AcMortM, r.AcSelM, r.AcErrM,
                r.AcConsKgM, r.ConsAcGrM, r.GrAveDiaM,
                r.DifConsMPct, r.DifPesoMPct, r.RetiroMPct, r.RetiroMAcPct,
                // Rel/Guías
                r.RelMHPct,
                r.PesoHGuia, r.UnifHGuia, r.ConsAcGrHGuia, r.GrAveDiaHGuia, r.MortHPctGuia,
                r.PesoMGuia, r.UnifMGuia, r.ConsAcGrMGuia, r.GrAveDiaMGuia, r.MortMPctGuia,
                r.AlimentoHGuia, r.AlimentoMGuia
            ))
            .ToListAsync();

        return new ResultadoLevanteResponse(
            loteId, desde?.Date, hasta?.Date, items.Count, items);
    }

}

// (Opcional): Si tienes provider que acepta string GalponId
public interface IGramajeProviderV2
{
    Task<double?> GetGramajeGrPorAveAsync(string galponId, int semana, string tipoAlimento);
}

File: frontend/src/app/features/lote-levante/services/seguimiento-lote-levante.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable, map } from 'rxjs';
import { environment } from '../../../../environments/environment';

export interface SeguimientoLoteLevanteDto {
  id: number;
  fechaRegistro: string;          // ISO
  loteId: string;

  mortalidadHembras: number;
  mortalidadMachos: number;
  selH: number;
  selM: number;
  errorSexajeHembras: number;
  errorSexajeMachos: number;

  tipoAlimento: string;
  consumoKgHembras: number;

  // Opcionales nuevos
  consumoKgMachos?: number | null;
  pesoPromH?: number | null;
  pesoPromM?: number | null;
  uniformidadH?: number | null;
  uniformidadM?: number | null;
  cvH?: number | null;
  cvM?: number | null;

  observaciones?: string;
  kcalAlH?: number | null;
  protAlH?: number | null;
  kcalAveH?: number | null;
  protAveH?: number | null;
  ciclo: string;                  // "Normal" | "Reforzado"
  tipoAlimentoHembras?: number | null; // (calculo interno, no se envía en create/update)
  tipoAlimentoMachos?: number | null;  // (calculo interno, no se envía en create/update)
}

export interface CreateSeguimientoLoteLevanteDto {
  fechaRegistro: string;          // ISO
  loteId: string;

  mortalidadHembras: number;
  mortalidadMachos: number;
  selH: number;
  selM: number;
  errorSexajeHembras: number;
  errorSexajeMachos: number;

  tipoAlimento: string;
  consumoKgHembras: number;

  // Opcionales nuevos
  consumoKgMachos?: number | null;
  pesoPromH?: number | null;
  pesoPromM?: number | null;
  uniformidadH?: number | null;
  uniformidadM?: number | null;
  cvH?: number | null;
  cvM?: number | null;

  observaciones?: string;
  kcalAlH?: number | null;
  protAlH?: number | null;
  kcalAveH?: number | null;
  protAveH?: number | null;
  ciclo: string;
  tipoAlimentoHembras?: number | null; // (calculo interno, no se envía en create/update)
  tipoAlimentoMachos?: number | null;  // (calculo interno, no se envía en create/update)
}

export interface UpdateSeguimientoLoteLevanteDto extends CreateSeguimientoLoteLevanteDto {
  id: number;
}

export interface ResultadoLevanteItemDto {
  fecha: string;            // "2025-09-08T00:00:00"
  edadDias: number | null;  // Cambiado de edadSemana a edadDias
  edadSemana?: number | null; // @deprecated - mantener para compatibilidad

  hembraViva: number | null;
  mortH: number; selH: number; errH: number;
  consKgH: number | null; pesoH: number | null; unifH: number | null; cvH: number | null;
  mortHPct: number | null; selHPct: number | null; errHPct: number | null;
  msEhH: number | null; acMortH: number | null; acSelH: number | null; acErrH: number | null;
  acConsKgH: number | null; consAcGrH: number | null; grAveDiaH: number | null;
  difConsHPct: number | null; difPesoHPct: number | null; retiroHPct: number | null; retiroHAcPct: number | null;

  machoVivo: number | null;
  mortM: number; selM: number; errM: number;
  consKgM: number | null; pesoM: number | null; unifM: number | null; cvM: number | null;
  mortMPct: number | null; selMPct: number | null; errMPct: number | null;
  msEmM: number | null; acMortM: number | null; acSelM: number | null; acErrM: number | null;
  acConsKgM: number | null; consAcGrM: number | null; grAveDiaM: number | null;
  difConsMPct: number | null; difPesoMPct: number | null; retiroMPct: number | null; retiroMAcPct: number | null;

  relMHPct: number | null;

  pesoHGuia: number | null; unifHGuia: number | null; consAcGrHGuia: number | null; grAveDiaHGuia: number | null; mortHPctGuia: number | null;
  pesoMGuia: number | null; unifMGuia: number | null; consAcGrMGuia: number | null; grAveDiaMGuia: number | null; mortMPctGuia: number | null;
  alimentoHGuia: string | null; alimentoMGuia: string | null;
}

export interface ResultadoLevanteResponse {
  loteId: string;
  desde: string | null;
  hasta: string | null;
  total: number;
  items: ResultadoLevanteItemDto[];
}

@Injectable({ providedIn: 'root' })
export class SeguimientoLoteLevanteService {
  /** Nota: environment.apiUrl debe incluir `/api` (ej: http://localhost:5002/api) */
  private readonly baseUrl = `${environment.apiUrl}/SeguimientoLoteLevante`;

  constructor(private http: HttpClient) {}

  /** GET general usando el endpoint de filtro sin parámetros */
  getAll(): Observable<SeguimientoLoteLevanteDto[]> {
    return this.http.get<SeguimientoLoteLevanteDto[]>(`${this.baseUrl}/filtro`);
  }

  /**
   * Polyfill de GetById: trae todo y filtra.
   * (Si el backend agrega GET /{id}, cámbialo por una llamada directa.)
   */
  getById(id: number): Observable<SeguimientoLoteLevanteDto> {
    return this.getAll().pipe(
      map(list => {
        const found = (list ?? []).find(x => x.id === id);
        if (!found) throw new Error(`Seguimiento ${id} no encontrado`);
        return found;
      })
    );
  }

  /** GET por LoteId */
  getByLoteId(loteId: number): Observable<SeguimientoLoteLevanteDto[]> {  // Changed from string to number
    return this.http.get<SeguimientoLoteLevanteDto[]>(
      `${this.baseUrl}/por-lote/${encodeURIComponent(loteId.toString())}`  // Convert to string for URL
    );
  }

  /** Filtro (loteId, desde, hasta) en ISO */
  filter(params: { loteId?: string; desde?: string | Date; hasta?: string | Date }): Observable<SeguimientoLoteLevanteDto[]> {
    let hp = new HttpParams();
    if (params.loteId) hp = hp.set('loteId', params.loteId);
    if (params.desde)  hp = hp.set('desde', this.toIso(params.desde));
    if (params.hasta)  hp = hp.set('hasta', this.toIso(params.hasta));
    return this.http.get<SeguimientoLoteLevanteDto[]>(`${this.baseUrl}/filtro`, { params: hp });
  }

  /** Crear */
  create(dto: CreateSeguimientoLoteLevanteDto): Observable<SeguimientoLoteLevanteDto> {
    return this.http.post<SeguimientoLoteLevanteDto>(this.baseUrl, dto);
  }

  /** Actualizar */
  update(dto: UpdateSeguimientoLoteLevanteDto): Observable<SeguimientoLoteLevanteDto> {
    return this.http.put<SeguimientoLoteLevanteDto>(`${this.baseUrl}/${dto.id}`, dto);
  }

  /** Eliminar */
  delete(id: number): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl}/${id}`);
  }

  // Helpers
  private toIso(d: string | Date): string {
    const dd = typeof d === 'string' ? new Date(d) : d;
    return dd.toISOString();
    // Si el back requiriera fecha sin hora: return dd.toISOString().substring(0, 10);
  }

  getResultado(params: {
    loteId: number;  // Changed from string to number
    desde?: string | Date;
    hasta?: string | Date;
    recalcular?: boolean;     // default true
  }): Observable<ResultadoLevanteResponse> {
    const { loteId } = params;
    let hp = new HttpParams()
      .set('recalcular', String(params.recalcular ?? true));
    if (params.desde) hp = hp.set('desde', this.toIso(params.desde));
    if (params.hasta) hp = hp.set('hasta', this.toIso(params.hasta));
    const url = `${this.baseUrl}/por-lote/${encodeURIComponent(loteId)}/resultado`;
    return this.http.get<ResultadoLevanteResponse>(url, { params: hp });
  }

 
}
